packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -198,6 +198,162 @@ class RenderEditable extends RenderBox {
+    final int pressedKeyMetaState = rawAndroidEvent.metaState;
+
+    if (selection.isCollapsed) {
+      _extentOffset = selection.extentOffset;
+      _baseOffset = selection.baseOffset;
+    }
+
+    // Update current key states
+    final bool shift = pressedKeyMetaState & _kShiftMask > 0;
+    final bool ctrl = pressedKeyMetaState & _kControlMask > 0;
+
+    final bool rightArrow = pressedKeyCode == _kRightArrowCode;
+    final bool leftArrow = pressedKeyCode == _kLeftArrowCode;
+    final bool upArrow = pressedKeyCode == _kUpArrowCode;
+    final bool downArrow = pressedKeyCode == _kDownArrowCode;
+    final bool arrow = leftArrow || rightArrow || upArrow || downArrow;
+
+    // We will only move select or more the caret if an arrow is pressed
+    if (arrow) {
+      int newOffset = _extentOffset;
+
+      // Because the user can use multiple keys to change how he selects
+      // the new offset variable is threaded through these four functions
+      // and potentially changes after each one.
+      newOffset = _handleControl(rightArrow, leftArrow, ctrl, newOffset);
+      newOffset = _handleHorizontalArrows(rightArrow, leftArrow, shift, newOffset);
+      newOffset = _handleVerticalArrows(upArrow, downArrow, shift, newOffset);
+      newOffset = _handleShift(rightArrow, leftArrow, shift, newOffset);
+
+      _extentOffset = newOffset;
+    }
+  }
+
+  // Handles the final selection of text or removal of the selection and placing
+  // of the caret
+  int _handleShift(bool rightArrow, bool leftArrow, bool shift, int newOffset) {
+    // For some reason, deletion only works if the base offset is less
+    // than the extent offset.
+    if (shift) {
+      if (_baseOffset < newOffset)
+        onSelectionChanged(new TextSelection(
+          baseOffset: _baseOffset, extentOffset: newOffset),
+          this, SelectionChangedCause.keyboard);
+      else
+        onSelectionChanged(new TextSelection(
+          baseOffset: newOffset , extentOffset: _baseOffset),
+          this, SelectionChangedCause.keyboard);
+    }
+    // We want to put the cursor at the correct location depending on which
+    // arrow is used while there is a selection.
+    else {
+      if (!selection.isCollapsed) {
+        if (leftArrow)
+          newOffset = _baseOffset < _extentOffset ? _baseOffset : _extentOffset;
+        else if (rightArrow)
+          newOffset = _baseOffset > _extentOffset ? _baseOffset : _extentOffset;
+      }
+      onSelectionChanged(new TextSelection.fromPosition(
+        new TextPosition(offset: newOffset)),
+        this, SelectionChangedCause.keyboard);
+    }
+    return newOffset;
+  }
+
+  // Handles moving the cursor vertically as well as taking care of the
+  // case where the user moves the cursor to the end or beginning of the text
+  // and then back up or down.
+  int _handleVerticalArrows(bool upArrow, bool downArrow, bool shift, int newOffset) {
+    if (downArrow || upArrow) {
```

Comment:
as with ctrl, there's probably no point calling this method at all if neither is true

Reply:
I mean, that's true for every button? I feel like it's cleaner to check for the key presses within the function though I can move them outside if you feel that would be better.

Reply:
It doesn't make a huge difference in this case, but in the general case you want code to only call methods that they want to actually run, because calling a method itself has a non-trivial cost.